### PR TITLE
GDPR - Erasure and PII cleanup

### PR DESCRIPTION
**Pull request by mikejolley**
This PR implements erasure requests in WP 4.9.6. This may need some tweaks when that is finalised, but the majority of the work is now finished.

Changes:

- `abstract-wc-privacy.php` file which extendables want to extend for extensions. This takes care of common methods and privacy notices. Exporter and erasure code has been updated/refactored to use this new class.
- Settings:
   - Account erasure requests - control what data gets deleted when a user is removed. You can choose to remove order data too, or retain it.
   - Personal data retention - Choose how long accounts and orders are kept before they are deleted or anonymized. This relies on cron so be sure to deactivate and reactivate WooCommerce if testing.
- Last active - On each page view, on login, and order creating an order, user meta tracks when an account was last active. This aids cleanup.
- Anonymize bulk action - available on the order screen. Will remove user data and order notes.

Todos, which can be done after merge:

- Get final approval on wording and texts used throughout from legal team
- Ensure the final erasure process in WP core works with this code and update the links to those screens we have in our own settings pages.

### CODE REVIEW

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Some time ago we agreed to avoid TitleCase on WooCommerce, but this seems like is introducing again a lot of strings with TitleCase.

**Comment by mikejolley** [comment]
@claudiosanches Customizer is not part of WooCommerce so if it uses Title case shouldn't we too? Otherwise it's not consistent.

**Comment by claudiosanches** [comment]
@mikejolley we stopped doing it even in other places on admin.

**Review by claudiosanches — APPROVED** [review]

File: includes/admin/settings/class-wc-settings-accounts.php [hunk 1]
```diff
@@ -133,50 +154,65 @@ public function get_settings() {
...
+					'placeholder' => __( 'N/A', 'woocommerce' ),
+					'default'     => array(
+						'number' => '',
+						'unit'   => 'months',
+					),
+					'autoload'    => false,
+				),
+				array(
+					'title'       => __( 'Retain pending orders ', 'woocommerce' ),
+					'desc_tip'    => __( 'Pending orders are unpaid and may have been abadonned by the customer. They will be trashed after the specified duration.', 'woocommerce' ),
```
Comment (claudiulodro) [inline includes/admin/settings/class-wc-settings-accounts.php]: `abadonned`. The tip for the next setting has this spelling error also.

File: includes/class-wc-privacy-erasers.php [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+			$response['done'] = 10 > count( $orders );
+		} else {
+			$response['done'] = true;
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Finds and exports customer download logs by email address.
```
Comment (claudiulodro) [inline includes/class-wc-privacy-erasers.php]: This description seems to be incorrect.

File: includes/class-wc-privacy-erasers.php [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+		);
+
+		if ( $user instanceof WP_User ) {
+			$downloads_query['user_id'] = (int) $user->ID;
+		} else {
+			$downloads_query['user_email'] = $email_address;
+		}
+
+		$customer_download_data_store     = WC_Data_Store::load( 'customer-download' );
+		$customer_download_log_data_store = WC_Data_Store::load( 'customer-download-log' );
```
Comment (claudiulodro) [inline includes/class-wc-privacy-erasers.php]: `$customer_download_log_data_store` is not used anywhere in this method.

File: includes/class-wc-privacy-erasers.php [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	 * @param int    $page  Page.
+	 * @return array An array of personal data in name value pairs
+	 */
+	public static function customer_data_eraser( $email_address, $page ) {
+		$response = array(
+			'items_removed'  => false,
+			'items_retained' => false,
+			'messages'       => array(),
+			'done'           => true,
+		);
```
Comment (claudiulodro) [inline includes/class-wc-privacy-erasers.php]: I was not able to get the Remove Personal Data tool to work successfully. I get the following error every time:
`{"success":false,"data":"Error: Expected num_items_removed key in response array from Customer Data eraser (index 2)."}`

[It seems that the correct keys are num_items_removed and num_items_retained](https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/ajax-actions.php#L4586-L4603), and they expect numbers. Maybe we're on different versions of the WP branch? I tested with the latest SVN-synced version from GitHub.

File: includes/class-wc-privacy.php [hunk 5]
```diff
@@ -178,533 +223,92 @@ public static function anonymize_completed_orders( $limit = 20, $page = 1 ) {
...
+					'type'    => 'NUMERIC',
+				),
+				array(
+					'key'     => 'wc_last_active',
+					'value'   => '0',
+					'compare' => '>',
+					'type'    => 'NUMERIC',
+				),
+			),
+		) );
```
Comment (claudiulodro) [inline includes/class-wc-privacy.php]: If a user never logs in after a site is updated to 3.4 they will never get cleaned up by this. Just a heads-up so you're aware. 

I think the best solution if that issue needs to be addressed is to have the update script give all users a `wc_last_active` of the current time when the update script runs.

File: includes/class-wc-privacy-erasers.php [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+		if ( $user instanceof WP_User ) {
+			$downloads_query['user_id'] = (int) $user->ID;
+		} else {
+			$downloads_query['user_email'] = $email_address;
+		}
+
+		$customer_download_data_store     = WC_Data_Store::load( 'customer-download' );
+		$customer_download_log_data_store = WC_Data_Store::load( 'customer-download-log' );
+		$downloads                        = $customer_download_data_store->get_downloads( $downloads_query );
```
Comment (claudiulodro) [inline includes/class-wc-privacy-erasers.php]: `$downloads` is not used anywhere in this method.

Should this method also delete stuff from the download logs? I wasn't able to successfully run the removal tool (see comment below) but I don't see a mechanism for it here. It seems like we should delete download logs for the email (and user IP if possible).

screen shot 2018-04-30 at 3 12 35 pm

**Review by claudiulodro — CHANGES_REQUESTED** [review]
The features look like they will do the job well and overall this looks good, but there are a few bugs/tweaks that should be taken care of. Also merge conflicts need to be resolved. :)

File: includes/class-wc-privacy-erasers.php (hunk 4 above)
Comment (mikejolley) [reply includes/class-wc-privacy-erasers.php]: https://core.trac.wordpress.org/ticket/43602#comment:44

File: includes/class-wc-privacy-erasers.php (hunk 6 above)
Comment (mikejolley) [reply includes/class-wc-privacy-erasers.php]: This is managed by the DB schema now - we added some foreign keys.

**Comment by claudiulodro** [comment]
Didn't notice when I was reviewing this yesterday, but [a unit test is throwing fatal errors now](https://travis-ci.org/woocommerce/woocommerce/jobs/373461524) and should be updated.

File: includes/class-wc-privacy.php (hunk 5 above)
Comment (mikejolley) [reply includes/class-wc-privacy.php]: Good idea. Done this now.

**Comment by mikejolley** [comment]
Merging this for @bor0 now tests pass.

**Pull request merged**